kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java

```diff
@@ -0,0 +1,194 @@
+      os.close();
+    } catch (IOException e) {
+      throw new DatasetIOException("Could not access signal path: " + signalPath, e);
+    }
+  }
+
+  /**
+   * Check the last time the specified constraints have been signaled as ready.
+   *
+   * @param viewConstraints The constraints to check for a signal.
+   *
+   * @return the timestamp of the last time the constraints were signaled as ready.
+   *          if the constraints have never been signaled, -1 will be returned.
+   *
+   * @throws DatasetException if the signals could not be accessed.
+   */
+  public long getReadyTimestamp(Constraints viewConstraints) {
+    String normalizedConstraints = getNormalizedConstraints(viewConstraints);
+
+    Path signalPath = new Path(signalDirectory, normalizedConstraints);
+    // check if the signal exists
+    try {
+      try {
+        FileStatus signalStatus = rootFileSystem.getFileStatus(signalPath);
+        return signalStatus.getModificationTime();
+      } catch (final FileNotFoundException ex) {
+        // empty, will be thrown when the signal path doesn't exist
+      }
+      return -1;
+    } catch (NumberFormatException e) {
+      throw new DatasetException("Signal contents were modified to not be a number: "
+                                   + signalPath, e);
+    } catch (IOException e) {
+      throw new DatasetIOException("Could not access signal path: " + signalPath, e);
+    }
+  }
+
+  /**
+   * Get a normalized query string for the {@link Constraints} that identifies a
+   * logical {@code View}.
+   *
+   * The normalized constraints will match to the query portion of a URI that will
+   * be exactly the same as another logically equivalent URI.
+   * (where, for example, the query parameters may be re-ordered)
+   *
+   * If the constraints are {@link Constraints#isUnbounded() unbounded} a special case
+   * of "unbounded" will be returned.
+   *
+   * @return a normalized query string for the specified constraints.
+   *
+   * @since 1.1
+   */
+  public static String getNormalizedConstraints(Constraints constraints) {
+    // the constraints map isn't naturally ordered
+    // we want to ensure that our output is
+
+    if(constraints.isUnbounded()) {
+      // unbounded constrains is a special case, here we just use
+      // "unbounded" as the constraint
+      return UNBOUNDED_CONSTRAINT;
+    }
+
+    Map<String, String> orderedConstraints = constraints.toNormalizedQueryMap();
+
+    List<String> parts = new ArrayList<String>();
+    // build a query portion of the URI
+    for (Map.Entry<String, String> entry : orderedConstraints.entrySet()) {
+      StringBuilder builder = new StringBuilder();
+      String key = entry.getKey();
+      String value = entry.getValue();
+      builder.append(key);
+      builder.append("=");
+      if(value != null) {
+        builder.append(value);
+      }
+      parts.add(builder.toString());
+    }
+
+    return Joiner.on('&').join(parts);
```

Comment:
The '&' is safe for HDFS. Although is certainly awkward for shell usage I'm not sure how often you'd interact with these files through a shell. I added another test for signal manager with multiple constraints here: ecb9065. Additionally the '&' mirrors the query portion of the view URI so it feels like the most natural fit.

Covered-by, and contained-in views would require that we be able to convert from the ready signal files to the constraints, and not having to read the blocks to get the contents of hashed file names to get the constraints would speed up loading for that use case.

We have a stop gap solution we were thinking of using in house before we go to Oozie 4 to use the signal files themselves for our input events. This would require the file names to be predictable and simple and a hash would throw a bit of a wrench in that idea. I wouldn't add to much weight to this for deciding if we should go with hashes or not, but it is a good example of how a creative consumer might get some use out of readable file names. (in addition to easier debugging)